# Multiple Linear Regression
Multiple independent variables in Linear Regression

In [1]:
from sklearn.linear_model import LinearRegression 
import pandas as pd

In [2]:
# Read csv file
car = pd.read_csv('./files/auto-mpg.csv', header=None, names=['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model year', 'origin', 'name'])
car.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   
3  16.0          8         304.0      150.0  3433.0          12.0          70   
4  17.0          8         302.0      140.0  3449.0          10.5          70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [3]:
# Set x, y data
x = car[['weight', 'cylinders']]
y = car[['mpg']]

In [4]:
# Split train & test data
from sklearn.model_selection import train_test_split

In [5]:
# Confirm split data
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size=0.85, random_state=0)
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((338, 2), (60, 2), (338, 1), (60, 1))

In [6]:
# Set model object
lm = LinearRegression()

In [7]:
# Fit data
lm.fit(x_train, y_train)

LinearRegression()

In [8]:
# Predict y_train_hat with linear model already fitted
y_train_hat = lm.predict(x_train)
y_train_hat[:5]

array([[19.31457821],
       [27.04576281],
       [24.18176499],
       [23.90895452],
       [12.1042136 ]])

In [9]:
# Compare with actual y_train value
y_train.iloc[:5]

mpg
200  18.0
206  26.5
107  18.0
15   22.0
106  12.0

In [10]:
# Get coefficient & intercept
lm.coef_, lm.intercept_

(array([[-0.00620024, -0.73757226]]), array([45.89966215]))

Above outcome means as follows:
* **y (mpg) = - 0.0620024 * (weight) - 0.73757226 * (cylinder) + 45.89966215** 

In [11]:
# R-squared for train data
lm.score(x_train, y_train)

0.6897269031891438

In [12]:
# R-squared for whole data with already fitted linear model
# R-squared of self.predict(x) and y
lm.score(x, y)

0.6960790024886645

In [13]:
# R-squared for test data with already fitted linear model
lm.score(x_test, y_test)

0.7259419804282132